File: py/road_hmm.py
"""Online HMM map matching (Newson & Krumm 2009, run as a causal forward filter).

Phase 9 alternative to the greedy / fixed-lag matchers in road_window.py. States are
directed road segments (consecutive OSM vertices) near the filter's position:

  emission    N(distance; 0, sigma) * N(heading error; 0, 25 deg)   sigma = max(4 m, filter sigma)
  transition  exp(-|route distance - distance driven| / beta)        route distance by Dijkstra
                                                                     on the directed road graph

It runs at 1 Hz in GNSS mode too, so it is already locked onto the right road when
an outage starts. The posterior of the best road is the match confidence; the
engine only uses a match that is confident AND made while the filter itself is
still tight (edge_engine._map_step_hmm). Roads can be restricted by OSM class:
service roads (car parks, driveways) are half of the IO-VNBD-area network and are
the parallel "roads" a drifting estimate snaps onto.
"""
from __future__ import annotations
import heapq, math
from collections import defaultdict
import numpy as np

R_EARTH = 6_371_000.0
CELL_M = 100.0
# carriageway half-width (m) by OSM highway class; the cross-track sigma is 0.6 x this
HALF_WIDTH = {"motorway": 7.0, "trunk": 7.0, "primary": 6.0, "secondary": 5.0, "tertiary": 4.5,
              "unclassified": 3.5, "residential": 3.5, "living_street": 3.0, "service": 3.0}
_GRAPHS = {}


class RoadGraph:
    """Directed segment graph of roads.bin ways in an ENU frame at (lat0, lon0)."""

    def __init__(self, ways, lat0, lon0, exclude=()):
        k = math.pi / 180 * R_EARTH; cl = math.cos(math.radians(lat0))
        keep = [w for w in ways if (w[4] if len(w) > 4 else None) not in exclude]
        la = np.concatenate([w[0] for w in keep]); lo = np.concatenate([w[1] for w in keep])
        wid = np.repeat(np.arange(len(keep)), [len(w[0]) for w in keep])
        key = np.stack([np.round(la * 1e7), np.round(lo * 1e7)], 1).astype(np.int64)
        _, node = np.unique(key, axis=0, return_inverse=True)
        node = node.ravel()
        e = (lo - lon0) * cl * k; n = (la - lat0) * k
        s = np.nonzero(wid[:-1] == wid[1:])[0]                  # point s -> s+1 is a segment
        s = s[(e[s + 1] != e[s]) | (n[s + 1] != n[s])]
        self.a, self.b = node[s], node[s + 1]
        self.p = np.stack([e[s], n[s]], 1)
        self.d = np.stack([e[s + 1] - e[s], n[s + 1] - n[s]], 1)
        self.len = np.hypot(self.d[:, 0], self.d[:, 1])
        self.brg = np.arctan2(self.d[:, 0], self.d[:, 1])       # compass bearing, geometry order
        w = wid[s]
        self.oneway = np.array([keep[i][3] for i in w], bool)
        self.tunnel = np.array([keep[i][2] for i in w], bool)
        cls = [keep[i][4] if len(keep[i]) > 4 else None for i in w]
        hw = {c: HALF_WIDTH.get((c or "").replace("_link", ""), 3.5) for c in set(cls)}
        self.half_width = np.array([hw[c] for c in cls])
        # directed adjacency: node -> [(next node, seg, dirn)]
        self.out = defaultdict(list)
        nbrs = defaultdict(set)
        for i, (u, v) in enumerate(zip(self.a.tolist(), self.b.tolist())):
            self.out[u].append((v, i, 1))
            if not self.oneway[i]:
                self.out[v].append((u, i, -1))
            nbrs[u].add(v); nbrs[v].add(u)
        self.degree = defaultdict(int, {u: len(x) for u, x in nbrs.items()})
        # uniform grid over segment bounding boxes
        lo_c = np.floor(np.minimum(self.p, self.p + self.d) / CELL_M).astype(int)
        hi_c = np.floor(np.maximum(self.p, self.p + self.d) / CELL_M).astype(int)
        grid = defaultdict(list)
        for i in range(len(self.len)):
            for cx in range(lo_c[i, 0], hi_c[i, 0] + 1):
                for cy in range(lo_c[i, 1], hi_c[i, 1] + 1):
                    grid[(cx, cy)].append(i)
        self.grid = {c: np.array(v) for c, v in grid.items()}

    @classmethod
    def cached(cls, ways, lat0, lon0, exclude=()):
        key = (id(ways), round(lat0, 7), round(lon0, 7), tuple(sorted(exclude)))
        if key not in _GRAPHS:
            _GRAPHS[key] = cls(ways, lat0, lon0, exclude)
        return _GRAPHS[key]

    def candidates(self, xy, r):
        """-> (seg ids, foot points, distance, t along segment) within r of xy."""
        c0 = np.floor((np.asarray(xy) - r) / CELL_M).astype(int)
        c1 = np.floor((np.asarray(xy) + r) / CELL_M).astype(int)
        ids = [self.grid[(cx, cy)] for cx in range(c0[0], c1[0] + 1) for cy in range(c0[1], c1[1] + 1)
               if (cx, cy) in self.grid]
        if not ids:
            return (np.empty(0, int),) * 4
        ids = np.unique(np.concatenate(ids))
        p, d, L = self.p[ids], self.d[ids], self.len[ids]
        t = np.clip(((xy[0] - p[:, 0]) * d[:, 0] + (xy[1] - p[:, 1]) * d[:, 1]) / (L * L), 0.0, 1.0)
        foot = p + t[:, None] * d
        dist = np.hypot(foot[:, 0] - xy[0], foot[:, 1] - xy[1])
        m = dist <= r
        return ids[m], foot[m], dist[m], t[m]

    def reach(self, src, max_d):
        """Bounded Dijkstra on the directed graph: node -> driving distance from src."""
        dist = {src: 0.0}; pq = [(0.0, src)]
        while pq:
            dd, u = heapq.heappop(pq)
            if dd > dist.get(u, math.inf) or dd > max_d:
                continue
            for v, s, _ in self.out.get(u, ()):
                nd = dd + self.len[s]
                if nd < dist.get(v, math.inf) and nd <= max_d:
                    dist[v] = nd; heapq.heappush(pq, (nd, v))
        return dist


class HMMMatcher:
    def __init__(self, graph, sigma_min=4.0, beta=8.0, radius=45.0, heading_sigma_deg=25.0,
                 max_cands=10):
        self.g = graph
        self.sigma_min, self.beta, self.radius = sigma_min, beta, radius
        self.hs = math.radians(heading_sigma_deg)
        self.max_cands = max_cands
        self.reset()

    def reset(self):
        self.prev = None; self.logp = None; self.travel = 0.0; self.last = None

    def _cands(self, xy, psi, moving, r):
        ids, foot, dist, t = self.g.candidates(xy, r)
        if len(ids) == 0:
            return None
        dirn = np.ones(len(ids), int)
        head = self.g.brg[ids].copy()
        if moving:
            rev = ~self.g.oneway[ids] & (np.abs(_wrap_arr(psi - head)) > math.pi / 2)
            dirn[rev] = -1; head[rev] = _wrap_arr(head[rev] + math.pi)
        o = np.argsort(dist)[: self.max_cands]
        return dict(seg=ids[o], dirn=dirn[o], foot=foot[o], dist=dist[o], t=t[o], head=head[o])

    def _route(self, pv, i, cu, j, cache):
        s0, d0, t0 = int(pv["seg"][i]), int(pv["dirn"][i]), pv["t"][i]
        s1, d1, t1 = int(cu["seg"][j]), int(cu["dirn"][j]), cu["t"][j]
        L0, L1 = self.g.len[s0], self.g.len[s1]
        if s0 == s1 and d0 == d1:
            along = (t1 - t0) * L0 * d0
            return along if along >= -3.0 else math.inf
        exit0 = int(self.g.b[s0] if d0 > 0 else self.g.a[s0])
        if (s0, d0) not in cache:
            cache[(s0, d0)] = self.g.reach(exit0, self.travel + 60.0)
        entry1 = int(self.g.a[s1] if d1 > 0 else self.g.b[s1])
        sp = cache[(s0, d0)].get(entry1)
        if sp is None:
            return math.inf
        return ((1 - t0) * L0 if d0 > 0 else t0 * L0) + sp + (t1 * L1 if d1 > 0 else (1 - t1) * L1)

    def update(self, xy, psi, speed, pos_sigma):
        """One 1 Hz step at the filter's (e, n, psi). Call add_travel() in between."""
        moving = speed > 2.0
        sig = max(self.sigma_min, pos_sigma)
        cu = self._cands(np.asarray(xy, float), psi, moving, max(self.radius, 3 * sig))
        if cu is None:
            self.reset(); return None
        em = -0.5 * (cu["dist"] / sig) ** 2
        if moving:
            em = em - 0.5 * (_wrap_arr(psi - cu["head"]) / self.hs) ** 2
        if self.prev is None:
            lp = em
        else:
            pv, plp, cache = self.prev, self.logp, {}
            lp = np.full(len(em), -np.inf)
            for j in range(len(em)):
                best = -np.inf
                for i in range(len(plp)):
                    if plp[i] == -np.inf:
                        continue
                    rd = self._route(pv, i, cu, j, cache)
                    if rd < math.inf:
                        best = max(best, plp[i] - abs(rd - self.travel) / self.beta)
                lp[j] = best + em[j]
            if not np.isfinite(lp).any():               # broken chain (off-map, big jump): restart
                lp = em
        lp = lp - lp.max()
        post = np.exp(lp); post /= post.sum()
        self.prev, self.logp, self.travel = cu, np.log(np.maximum(post, 1e-300)), 0.0
        k = int(np.argmax(post)); s = int(cu["seg"][k])
        L = self.g.len[s]; tk = cu["t"][k]
        # Confidence in the ROAD, not the segment: OSM draws a road as many short
        # vertex-to-vertex segments, and near a vertex two of them are equally likely.
        # Sum the posterior of candidates on the same line (feet within 5 m, same
        # direction within 30 deg). A parallel street or the opposite carriageway is
        # further than 5 m away, so it still counts against the match.
        df = np.hypot(*(cu["foot"] - cu["foot"][k]).T)
        same = (df < 5.0) & (np.abs(_wrap_arr(cu["head"] - cu["head"][k])) < math.radians(30))
        self.last = dict(seg=s, foot=cu["foot"][k], bearing=float(cu["head"][k]), dist=float(cu["dist"][k]),
                         conf=float(post[same].sum()), seg_conf=float(post[k]),
                         seg_len=float(L), end_dist=float(min(tk, 1 - tk) * L),
                         half_width=float(self.g.half_width[s]), tunnel=bool(self.g.tunnel[s]))
        return self.last

    def add_travel(self, d):
        self.travel += d


def _wrap_arr(a):
    return (np.asarray(a) + np.pi) % (2 * np.pi) - np.pi

File: android/app/src/main/java/com/offmaps/nav/RoadHmm.kt
package com.offmaps.nav

import java.util.PriorityQueue
import kotlin.math.PI
import kotlin.math.abs
import kotlin.math.atan2
import kotlin.math.exp
import kotlin.math.floor
import kotlin.math.hypot
import kotlin.math.ln
import kotlin.math.max
import kotlin.math.min

/**
 * Online HMM road matching (Newson & Krumm 2009 as a causal forward filter): the port of
 * py/road_hmm.py, checked against it by py/tests/test_kotlin_ports.py (tools/kthost).
 * Pure Kotlin, no Android deps. Phase 9 measured it on real IO-VNBD roads: train drives
 * 13.6 -> 11.7 % mean median drift, where the greedy RoadMatcher made outages worse
 * (py/phase9_map_eval.py, out/phase9/).
 *
 * States are directed road segments (consecutive vertices of a way) near the filter's
 * position; once a second:
 *   emission    N(distance; 0, sigma) * N(heading error; 0, 25 deg), sigma = max(4 m, filter sigma)
 *   transition  exp(-|route distance - distance driven| / beta), route distance by bounded
 *               Dijkstra on the directed (one-way aware) road graph
 * [update] returns the best segment with a confidence per ROAD (posterior of the candidates
 * on the same line), since OSM splits one road into many short segments.
 *
 * The phone cannot hold a city's graph, so a [WINDOW_M] window of ways is rebuilt whenever
 * the car has moved [RECENTER_M]. Candidates carry stable keys (way, vertex) and graph
 * nodes are keyed by their exact 1e-7 degree coordinates, so the HMM state survives a
 * rebuild and the result equals the Python reference on the whole network: candidates
 * lie within RADIUS_MAX of the car and routes within (distance driven + 60 m) of them,
 * both well inside the window.
 */
interface RoadSource {
    fun size(w: Int): Int
    fun lat7(w: Int, i: Int): Int
    fun lon7(w: Int, i: Int): Int
    fun tunnel(w: Int): Int
    fun oneway(w: Int): Int
    fun hwCode(w: Int): Int                    // tools/osm_layers.py HW_CODE, 0 = unknown
    fun waysNear(lat: Double, lon: Double, radiusM: Double): IntArray
}

class RoadHmm(private val roads: RoadSource, private val enu: Enu,
              private val exclude: Set<Int> = setOf(HW_SERVICE)) {

    class Match(val way: Int, val vtx: Int, val footE: Double, val footN: Double, val bearing: Double,
                val dist: Double, val conf: Double, val segConf: Double, val segLen: Double,
                val endDist: Double, val halfWidth: Double, val tunnel: Boolean)

    // ---- the window graph ----
    private var nSeg = 0
    private var segKey = LongArray(0)          // (way shl 8) or vtx: ascending == Python segment id order
    private var aNode = LongArray(0); private var bNode = LongArray(0)
    private var px = DoubleArray(0); private var py = DoubleArray(0)
    private var dx = DoubleArray(0); private var dy = DoubleArray(0)
    private var len = DoubleArray(0); private var brg = DoubleArray(0)
    private var oneway = BooleanArray(0); private var tun = BooleanArray(0); private var hw = DoubleArray(0)
    private var out = HashMap<Long, MutableList<IntArray>>()   // node -> [(seg, dirn)]; next node from seg
    private var grid = HashMap<Long, IntArray>()
    private var cE = 0.0; private var cN = 0.0; private var built = false
    var windowSegments = 0; private set

    // ---- HMM state: previous candidates (by value, so a rebuild doesn't break them) ----
    private class Cands(val key: LongArray, val dirn: IntArray, val t: DoubleArray, val exitNode: LongArray,
                        val entryNode: LongArray, val len: DoubleArray, val foot: Array<DoubleArray>,
                        val dist: DoubleArray, val head: DoubleArray, val seg: IntArray)
    private var prev: Cands? = null
    private var logp = DoubleArray(0)
    private var travel = 0.0
    var last: Match? = null; private set

    fun reset() { prev = null; logp = DoubleArray(0); travel = 0.0; last = null }
    fun addTravel(d: Double) { travel += d }

    private fun rebuild(e: Double, n: Double) {
        val ll = enu.toLatLon(e, n)
        val ids = roads.waysNear(ll[0], ll[1], WINDOW_M).filter { roads.hwCode(it) !in exclude }.sorted()
        var cap = 0
        for (w in ids) cap += roads.size(w) - 1
        val sk = LongArray(cap); val an = LongArray(cap); val bn = LongArray(cap)
        val sx = DoubleArray(cap); val sy = DoubleArray(cap); val ddx = DoubleArray(cap); val ddy = DoubleArray(cap)
        val ow = BooleanArray(cap); val tn = BooleanArray(cap); val hwd = DoubleArray(cap)
        var s = 0
        for (w in ids) {
            val k = roads.size(w)
            var prevKey = 0L; var pe = 0.0; var pn = 0.0
            for (i in 0 until k) {
                val la7 = roads.lat7(w, i); val lo7 = roads.lon7(w, i)
                val p = enu.toEn(la7 / SCALE, lo7 / SCALE)
                val key = (la7.toLong() shl 32) or (lo7.toLong() and 0xffffffffL)
                if (i > 0 && (p[0] != pe || p[1] != pn)) {
                    sk[s] = (w.toLong() shl 8) or (i - 1).toLong(); an[s] = prevKey; bn[s] = key
                    sx[s] = pe; sy[s] = pn; ddx[s] = p[0] - pe; ddy[s] = p[1] - pn
                    ow[s] = roads.oneway(w) == 1; tn[s] = roads.tunnel(w) == 1
                    hwd[s] = halfWidth(roads.hwCode(w)); s++
                }
                prevKey = key; pe = p[0]; pn = p[1]
            }
        }
        nSeg = s; segKey = sk.copyOf(s); aNode = an.copyOf(s); bNode = bn.copyOf(s)
        px = sx.copyOf(s); py = sy.copyOf(s); dx = ddx.copyOf(s); dy = ddy.copyOf(s)
        oneway = ow.copyOf(s); tun = tn.copyOf(s); hw = hwd.copyOf(s)
        len = DoubleArray(s) { hypot(dx[it], dy[it]) }
        brg = DoubleArray(s) { atan2(dx[it], dy[it]) }
        out = HashMap(2 * s)
        for (i in 0 until s) {
            out.getOrPut(aNode[i]) { ArrayList(3) }.add(intArrayOf(i, 1))
            if (!oneway[i]) out.getOrPut(bNode[i]) { ArrayList(3) }.add(intArrayOf(i, -1))
        }
        val g = HashMap<Long, MutableList<Int>>()
        for (i in 0 until s) {
            val x0 = floor(min(px[i], px[i] + dx[i]) / CELL_M).toInt(); val x1 = floor(max(px[i], px[i] + dx[i]) / CELL_M).toInt()
            val y0 = floor(min(py[i], py[i] + dy[i]) / CELL_M).toInt(); val y1 = floor(max(py[i], py[i] + dy[i]) / CELL_M).toInt()
            for (cx in x0..x1) for (cy in y0..y1) g.getOrPut(cell(cx, cy)) { ArrayList(4) }.add(i)
        }
        grid = HashMap(g.size * 2); for ((kk, v) in g) grid[kk] = v.toIntArray()
        cE = e; cN = n; built = true; windowSegments = s
    }

    private fun candidates(x: Double, y: Double, psi: Double, moving: Boolean, r: Double): Cands? {
        val ids = HashSet<Int>()
        for (cx in floor((x - r) / CELL_M).toInt()..floor((x + r) / CELL_M).toInt())
            for (cy in floor((y - r) / CELL_M).toInt()..floor((y + r) / CELL_M).toInt())
                grid[cell(cx, cy)]?.let { for (i in it) ids.add(i) }
        val c = ArrayList<DoubleArray>()                     // [seg, t, footE, footN, dist]
        for (i in ids) {
            val t = (((x - px[i]) * dx[i] + (y - py[i]) * dy[i]) / (len[i] * len[i])).coerceIn(0.0, 1.0)
            val fe = px[i] + t * dx[i]; val fn = py[i] + t * dy[i]
            val d = hypot(fe - x, fn - y)
            if (d <= r) c.add(doubleArrayOf(i.toDouble(), t, fe, fn, d))
        }
        if (c.isEmpty()) return null
        // by distance, ties by segment key (== np.lexsort((ids, dist)) in road_hmm.py)
        c.sortWith(compareBy<DoubleArray>({ it[4] }, { segKey[it[0].toInt()] }))
        val m = min(c.size, MAX_CANDS)
        val seg = IntArray(m) { c[it][0].toInt() }
        val dirn = IntArray(m) { 1 }; val head = DoubleArray(m) { brg[seg[it]] }
        if (moving) for (j in 0 until m) if (!oneway[seg[j]] && abs(wrap(psi - head[j])) > PI / 2) {
            dirn[j] = -1; head[j] = wrap(head[j] + PI)
        }
        return Cands(LongArray(m) { segKey[seg[it]] }, dirn, DoubleArray(m) { c[it][1] },
                     LongArray(m) { if (dirn[it] > 0) bNode[seg[it]] else aNode[seg[it]] },
                     LongArray(m) { if (dirn[it] > 0) aNode[seg[it]] else bNode[seg[it]] },
                     DoubleArray(m) { len[seg[it]] }, Array(m) { doubleArrayOf(c[it][2], c[it][3]) },
                     DoubleArray(m) { c[it][4] }, head, seg)
    }

    /** Bounded Dijkstra on the directed window graph: node -> driving distance from src. */
    private fun reach(src: Long, maxD: Double): HashMap<Long, Double> {
        val dist = HashMap<Long, Double>(); dist[src] = 0.0
        val pq = PriorityQueue<Pair<Double, Long>>(compareBy { it.first }); pq.add(0.0 to src)
        while (pq.isNotEmpty()) {
            val (dd, u) = pq.poll()
            if (dd > (dist[u] ?: Double.MAX_VALUE) || dd > maxD) continue
            for (edge in out[u] ?: continue) {
                val sg = edge[0]
                val v = if (edge[1] > 0) bNode[sg] else aNode[sg]
                val nd = dd + len[sg]
                if (nd < (dist[v] ?: Double.MAX_VALUE) && nd <= maxD) { dist[v] = nd; pq.add(nd to v) }
            }
        }
        return dist
    }

    private fun route(pv: Cands, i: Int, cu: Cands, j: Int, cache: HashMap<Int, HashMap<Long, Double>>): Double {
        val d0 = pv.dirn[i]; val t0 = pv.t[i]; val l0 = pv.len[i]
        val d1 = cu.dirn[j]; val t1 = cu.t[j]; val l1 = cu.len[j]
        if (pv.key[i] == cu.key[j] && d0 == d1) {
            val along = (t1 - t0) * l0 * d0
            return if (along >= -3.0) along else Double.POSITIVE_INFINITY
        }
        val r = cache.getOrPut(i) { reach(pv.exitNode[i], travel + 60.0) }
        val sp = r[cu.entryNode[j]] ?: return Double.POSITIVE_INFINITY
        return (if (d0 > 0) (1 - t0) * l0 else t0 * l0) + sp + (if (d1 > 0) t1 * l1 else (1 - t1) * l1)
    }

    /** One 1 Hz step at the filter's (e, n, psi); call [addTravel] with the distance driven in between. */
    fun update(e: Double, n: Double, psi: Double, speed: Double, posSigma: Double): Match? {
        if (!built || hypot(e - cE, n - cN) > RECENTER_M) rebuild(e, n)
        val moving = speed > 2.0
        val sig = max(SIGMA_MIN, posSigma)
        val cu = candidates(e, n, psi, moving, min(max(RADIUS, 3 * sig), RADIUS_MAX))
        if (cu == null) { reset(); return null }
        val m = cu.key.size
        val em = DoubleArray(m) {
            var v = -0.5 * (cu.dist[it] / sig) * (cu.dist[it] / sig)
            if (moving) { val dh = wrap(psi - cu.head[it]); v -= 0.5 * (dh / HEADING_SIGMA) * (dh / HEADING_SIGMA) }
            v
        }
        var lp: DoubleArray
        val pv = prev
        if (pv == null) lp = em
        else {
            val cache = HashMap<Int, HashMap<Long, Double>>()
            lp = DoubleArray(m) { Double.NEGATIVE_INFINITY }
            for (j in 0 until m) {
                var best = Double.NEGATIVE_INFINITY
                for (i in logp.indices) {
                    if (logp[i] == Double.NEGATIVE_INFINITY) continue
                    val rd = route(pv, i, cu, j, cache)
                    if (rd < Double.POSITIVE_INFINITY) best = max(best, logp[i] - abs(rd - travel) / BETA)
                }
                lp[j] = best + em[j]
            }
            if (lp.none { it.isFinite() }) lp = em              // broken chain (off-map, big jump): restart
        }
        val mx = lp.max()
        val post = DoubleArray(m) { exp(lp[it] - mx) }
        val sum = post.sum()
        for (i in 0 until m) post[i] /= sum
        prev = cu; logp = DoubleArray(m) { ln(max(post[it], 1e-300)) }; travel = 0.0
        var k = 0
        for (i in 1 until m) if (post[i] > post[k]) k = i         // first maximum, like np.argmax
        var conf = 0.0
        for (i in 0 until m) {
            val df = hypot(cu.foot[i][0] - cu.foot[k][0], cu.foot[i][1] - cu.foot[k][1])
            if (df < 5.0 && abs(wrap(cu.head[i] - cu.head[k])) < SAME_ROAD_DEG * PI / 180) conf += post[i]
        }
        val s = cu.seg[k]; val tk = cu.t[k]; val l = len[s]
        return Match((cu.key[k] shr 8).toInt(), (cu.key[k] and 0xff).toInt(), cu.foot[k][0], cu.foot[k][1],
                     cu.head[k], cu.dist[k], conf, post[k], l, min(tk, 1 - tk) * l, hw[s], tun[s]).also { last = it }
    }

    companion object {
        const val WINDOW_M = 1000.0
        const val RECENTER_M = 400.0
        const val CELL_M = 100.0
        const val SIGMA_MIN = 4.0
        const val BETA = 8.0
        const val RADIUS = 45.0
        const val RADIUS_MAX = 150.0
        const val MAX_CANDS = 10
        const val SAME_ROAD_DEG = 30.0
        val HEADING_SIGMA = 25.0 * PI / 180
        const val HW_SERVICE = 14                               // osm_layers.HW_CODE["service"]
        private const val SCALE = 1e7
        // osm_layers.ROAD_CLASS order (HW_CODE = index + 1); "_link" uses its road's width
        private val HW_NAMES = arrayOf("", "motorway", "motorway", "trunk", "trunk", "primary", "primary",
            "secondary", "secondary", "tertiary", "tertiary", "unclassified", "residential", "living_street", "service")
        private val HALF_WIDTH = mapOf("motorway" to 7.0, "trunk" to 7.0, "primary" to 6.0, "secondary" to 5.0,
            "tertiary" to 4.5, "unclassified" to 3.5, "residential" to 3.5, "living_street" to 3.0, "service" to 3.0)

        fun halfWidth(code: Int): Double = HALF_WIDTH[HW_NAMES.getOrElse(code) { "" }] ?: 3.5
        private fun cell(cx: Int, cy: Int) = (cx.toLong() shl 32) or (cy.toLong() and 0xffffffffL)

        /** Python's (a + pi) % 2pi - pi (floor modulo), so wraps agree to the last bit. */
        fun wrap(a: Double): Double {
            val m = (a + PI) % (2 * PI)
            return (if (m < 0) m + 2 * PI else m) - PI
        }
    }
}
